nvbench/detail/axes_iterator.cuh

```diff
@@ -0,0 +1,113 @@
+/*
+ *  Copyright 2021 NVIDIA Corporation
+ *
+ *  Licensed under the Apache License, Version 2.0 with the LLVM exception
+ *  (the "License"); you may not use this file except in compliance with
+ *  the License.
+ *
+ *  You may obtain a copy of the License at
+ *
+ *      http://llvm.org/foundation/relicensing/LICENSE.txt
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+#pragma once
+
+#include <nvbench/axis_base.cuh>
+#include <nvbench/type_axis.cuh>
+
+#include <functional>
+#include <utility>
+#include <vector>
+
+namespace nvbench
+{
+namespace detail
+{
+
+struct axis_index
+{
+  axis_index() = default;
+
+  explicit axis_index(const axis_base *axi)
+      : index(0)
+      , name(axi->get_name())
+      , type(axi->get_type())
+      , size(axi->get_size())
+      , active_size(axi->get_size())
+  {
+    if (type == nvbench::axis_type::type)
+    {
+      active_size =
+        static_cast<const nvbench::type_axis *>(axi)->get_active_count();
+    }
+  }
+  std::size_t index;
+  std::string name;
+  nvbench::axis_type type;
+  std::size_t size;
+  std::size_t active_size;
+};
+
+struct axis_space_iterator
```

Comment:
Given this is a pretty central piece of new machinery, this could use some docs for what is going on here.

Reply:
Also, given this isn't really an `iterator` in the sense that it doesn't satisfy any `std::iterator` concepts, I think the `iterator` name is confusing (it confused me anyways). 

I'd either make it satisfy an appropriate iterator concept or give it a different name.

Reply:
The class matches the naming pattern of the existing `state_iterator` ( https://github.com/NVIDIA/nvbench/blob/main/nvbench/detail/state_generator.cuh#L74 ) which calls the axis_iterator.